core-processor/src/ext.rs

```diff
@@ -319,6 +319,134 @@ impl GrowHandler for LazyGrowHandler {
+            }
+        }
+    }
+
+    fn reduce_gas(&mut self, limit: GasLimit) -> Result<(), FallibleExtError> {
+        if self.gas_counter.reduce(limit) == ChargeResult::NotEnough {
+            return Err(FallibleExecutionError::NotEnoughGas.into());
+        }
+
+        Ok(())
+    }
+
+    fn charge_message_value(&mut self, value: u128) -> Result<(), FallibleExtError> {
+        if self.value_counter.reduce(value) == ChargeResult::NotEnough {
+            return Err(FallibleExecutionError::NotEnoughValue.into());
+        }
+
+        Ok(())
+    }
+
+    fn mark_used(
+        &mut self,
+        ext: &Ext,
+        reservation_id: ReservationId,
+    ) -> Result<(), ReservationError> {
+        if let Some(
+            GasReservationState::Created { used, .. } | GasReservationState::Exists { used, .. },
+        ) = ext.context.gas_reserver.states().get(&reservation_id)
+        {
+            if *used {
+                Err(ReservationError::InvalidReservationId)
+            } else {
+                self.changes.push(Change::MarkUsed(reservation_id));
+                Ok(())
+            }
+        } else {
+            Err(ReservationError::InvalidReservationId)
+        }
+    }
+
+    fn gasless_with_delay(&mut self) -> Result<(), FallibleExtError> {
+        let prev_gasless_fee = self
+            .outgoing_glassless
+            .saturating_mul(self.mailbox_threshold);
+        self.reduce_gas(prev_gasless_fee)?;
+        self.outgoing_glassless = 1;
+        Ok(())
+    }
+
+    fn non_zero_or_delay_gasfull(&mut self) -> Result<(), FallibleExtError> {
+        let prev_gasless_fee = self
+            .outgoing_glassless
+            .saturating_mul(self.mailbox_threshold);
+        self.reduce_gas(prev_gasless_fee)?;
+        self.outgoing_glassless = 0;
+        Ok(())
+    }
+
+    fn charge_sending_fee(&mut self, ext: &mut Ext, delay: u32) -> Result<(), ChargeError> {
+        if delay == 0 {
+            self.charge_gas_if_enough(ext.context.message_context.settings().sending_fee)
+        } else {
+            self.charge_gas_if_enough(ext.context.message_context.settings().scheduled_sending_fee)
+        }
+    }
+
+    fn charge_gas_if_enough(&mut self, gas: u64) -> Result<(), ChargeError> {
+        if self.gas_counter.charge_if_enough(gas) == ChargeResult::NotEnough {
+            return Err(ChargeError::GasLimitExceeded);
+        }
+
+        if self.gas_allowance_counter.charge_if_enough(gas) == ChargeResult::NotEnough {
+            return Err(ChargeError::GasAllowanceExceeded);
+        }
+        Ok(())
+    }
+
+    fn apply(mut self, ext: &mut Ext) {
+        for change in self.changes.drain(..) {
+            match change {
+                Change::MarkUsed(id) => {
+                    assert!(ext.context.gas_reserver.mark_used(id).is_ok());
+                }
+            }
+        }
+
+        core::mem::swap(&mut ext.context.value_counter, &mut self.value_counter);
```

Comment:
why not to just `ext.context.value_counter = self.value_counter`

Reply:
Because we cannot move the `value_counter` out of `self`, at least rustc errors this

Reply:
Can leave it as is right now and in separate PR where we get rid of cloning counters it won't be a problem